x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -249,8 +251,20 @@ export class ContentStream extends Duplex {
     });
   }
 
-  private async flush(size = this.buffer.byteLength) {
-    const chunk = this.buffer.slice(0, size);
+  private async flush(size = this.currentByteLength) {
+    const buffersToConsume: Buffer[] = [];
+    let totalBytesConsumed = 0;
+
+    for (const buffer of this.buffers) {
+      if (totalBytesConsumed + buffer.byteLength <= size) {
+        buffersToConsume.push(buffer);
+        totalBytesConsumed += buffer.byteLength;
+      } else {
+        break;
+      }
+    }
+
+    const chunk = Buffer.concat(buffersToConsume);
```

Comment:
What if we optimize that even better? We can avoid walking through the buffers array.
```suggestion
  private async flush(size = this.bytesBuffered) {
    const buffer = Buffer.concat(this.buffers);
    const chunk = this.buffer.slice(0, size);
```

Reply:
looks like that was meant to be: `const chunk = buffer.slice(0, size);`

Reply:
I tested this out, and it looks like the bottleneck is created by a lot of GC activity. I think we should avoid copying data (Buffer.concat) as much as possible. From local testing the CPU gets flooded:

<img width="1242" alt="Screenshot 2021-12-06 at 10 24 35" src="https://user-images.githubusercontent.com/8155004/144823973-52731b16-2c85-4d18-9de0-2bb006722edd.png">

I think we should go with the current approach, happy for this to be revised at a later stage but we should test this code before making changes. See the comment I added here: https://github.com/elastic/kibana/pull/120365/commits/566296c69b56374dd638e7e59608b05a18e08b59

Reply:
What do you guys think?

Reply:
The comment makes sense. Impressive job on checking and debugging this 👏